File: scripts/deploy.js
require('dotenv').config();
const { ethers } = require('ethers');
const fs = require('fs');
const path = require('path');

async function main() {
  console.log('═══════════════════════════════════════════════════════');
  console.log('  COMFORTage T3.3 — Deploying DataIntegrity Contract');
  console.log('  Network: Reltime Mainnet (Chain ID: 32323)');
  console.log('═══════════════════════════════════════════════════════\n');

  // Setup provider and signer
  const rpcUrl = process.env.RELTIME_RPC_URL || 'https://mainnet.reltime.com/';
  const privateKey = process.env.DEPLOYER_PRIVATE_KEY;

  if (!privateKey) {
    throw new Error('DEPLOYER_PRIVATE_KEY not set in .env');
  }

  const provider = new ethers.JsonRpcProvider(rpcUrl, {
    name: 'reltime-mainnet',
    chainId: 32323,
  });

  const deployer = new ethers.Wallet(privateKey, provider);
  console.log("Deployer address:", deployer.address);

  const balance = await provider.getBalance(deployer.address);
  console.log("Deployer balance:", ethers.formatEther(balance), "ETH\n");

  // Verify network
  const network = await provider.getNetwork();
  console.log("Network Chain ID:", network.chainId);
  const blockNumber = await provider.getBlockNumber();
  console.log("Current block:", blockNumber, "\n");

  // Deploy
  console.log("Deploying DataIntegrity...");
  const artifactPath = path.join(__dirname, '../artifacts/contracts/DataIntegrity.sol/DataIntegrity.json');
  const artifact = JSON.parse(fs.readFileSync(artifactPath, 'utf8'));

  const ContractFactory = new ethers.ContractFactory(
    artifact.abi,
    artifact.bytecode,
    deployer
  );

  // Reltime has zero gas fees — use explicit gasPrice: 0
  const contract = await ContractFactory.deploy(deployer.address, {
    gasPrice: 0,
    gasLimit: 10000000,
  });

  await contract.waitForDeployment();
  const contractAddress = await contract.getAddress();

  console.log("\n✅ DataIntegrity deployed to:", contractAddress);
  console.log("\n──────────────────────────────────────────────────────");
  console.log("  IMPORTANT: Update your .env file:");
  console.log(`  CONTRACT_ADDRESS=${contractAddress}`);
  console.log("──────────────────────────────────────────────────────");

  // Grant roles to the DataIntegrityValidator service account
  // (You will set this address for T3.4's service account)
  const VALIDATOR_ADDRESS = process.env.VALIDATOR_ADDRESS;
  if (VALIDATOR_ADDRESS) {
    console.log("\nGranting VALIDATOR_ROLE to:", VALIDATOR_ADDRESS);
    const VALIDATOR_ROLE = await contract.VALIDATOR_ROLE();
    const tx = await contract.grantRole(VALIDATOR_ROLE, VALIDATOR_ADDRESS, {
      gasPrice: 0,
    });
    await tx.wait();
    console.log("✅ VALIDATOR_ROLE granted.");
  }

  const INGESTION_ADDRESS = process.env.INGESTION_ADDRESS;
  if (INGESTION_ADDRESS) {
    console.log("Granting INGESTION_ROLE to:", INGESTION_ADDRESS);
    const INGESTION_ROLE = await contract.INGESTION_ROLE();
    const tx = await contract.grantRole(INGESTION_ROLE, INGESTION_ADDRESS, {
      gasPrice: 0,
    });
    await tx.wait();
    console.log("✅ INGESTION_ROLE granted.");
  }

  console.log("\n🎉 Deployment complete!\n");
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error("❌ Deployment failed:", error);
    process.exit(1);
  });

File: deploy.py
#!/usr/bin/env python3

import json
import os
import sys
from web3 import Web3
from pathlib import Path
from dotenv import load_dotenv

# Load environment
load_dotenv()

print('═' * 59)
print('  COMFORTage T3.3 — Deploying DataIntegrity Contract')
print('  Network: Reltime Mainnet (Chain ID: 32323)')
print('═' * 59)
print()

try:
    # Setup Web3
    rpc_url = os.getenv('RELTIME_RPC_URL', 'https://mainnet.reltime.com/')
    w3 = Web3(Web3.HTTPProvider(rpc_url))
    
    if not w3.is_connected():
        raise Exception("Cannot connect to Reltime RPC")
    
    # Setup account
    private_key = os.getenv('DEPLOYER_PRIVATE_KEY')
    if not private_key:
        raise Exception("DEPLOYER_PRIVATE_KEY not set in .env")
    
    # Ensure key has 0x prefix
    if not private_key.startswith('0x'):
        private_key = '0x' + private_key
    
    account = w3.eth.account.from_key(private_key)
    print(f"Deployer address: {account.address}")
    
    # Check balance
    balance = w3.eth.get_balance(account.address)
    print(f"Deployer balance: {Web3.from_wei(balance, 'ether')} ETH\n")
    
    # Check network
    chain_id = w3.eth.chain_id
    print(f"Chain ID: {chain_id}")
    
    block_number = w3.eth.block_number
    print(f"Current block: {block_number}\n")
    
    # Load contract artifact
    artifact_path = Path(__file__).parent / 'artifacts' / 'contracts' / 'DataIntegrity.sol' / 'DataIntegrity.json'
    with open(artifact_path, 'r') as f:
        artifact = json.load(f)
    
    # Create contract factory
    Contract = w3.eth.contract(
        abi=artifact['abi'],
        bytecode=artifact['bytecode']
    )
    
    print("Deploying DataIntegrity...\n")
    
    # Build transaction
    constructor = Contract.constructor(account.address)
    tx = constructor.build_transaction({
        'from': account.address,
        'nonce': w3.eth.get_transaction_count(account.address),
        'gas': 10000000,
        'gasPrice': 0,  # Zero gas on Reltime PoA
    })
    
    # Sign and send
    signed_tx = w3.eth.account.sign_transaction(tx, private_key)
    tx_hash = w3.eth.send_raw_transaction(signed_tx.rawTransaction)
    
    print(f"Transaction hash: {tx_hash.hex()}")
    print("Waiting for receipt...\n")
    
    # Wait for receipt
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=300)
    contract_address = receipt['contractAddress']
    
    print(f"✅ DataIntegrity deployed to: {contract_address}\n")
    print('─' * 59)
    print("  IMPORTANT: Update your .env file:")
    print(f"  CONTRACT_ADDRESS={contract_address}")
    print('─' * 59)
    
    # Verify deployed
    code = w3.eth.get_code(contract_address)
    print(f"\n✅ Contract verified on chain (code length: {len(code)} bytes)")
    
    print("\n🎉 Deployment complete!\n")
    
except Exception as e:
    print(f"\n❌ Deployment failed: {str(e)}")
    sys.exit(1)
